cpp/src/arrow/columnar_format.cc

```diff
@@ -0,0 +1,1020 @@
+                       int16_t rep_level, int16_t def_level,
+                       MemoryPool* pool,
+                       std::unique_ptr<ShredNode>* node_out) {
+    std::unique_ptr<StructShredNode> node(new StructShredNode(field, pool));
+    node->rep_level_ = rep_level;
+    if (field->nullable()) {
+      def_level++;
+    }
+    node->def_level_ = def_level;
+    node->children_.resize(field->type()->num_children());
+    for (int i = 0; i < field->type()->num_children(); i++) {
+      std::unique_ptr<ShredNode> child;
+      RETURN_NOT_OK(CreateTree(field->type()->child(i), rep_level, def_level, pool, &child));
+      node->children_[i] = std::move(child);
+    }
+    *node_out = std::move(node);
+    return Status::OK();
+  }
+
+  Status Handle(const Array& array, int64_t i, int16_t rep_level) override {
+    return Handle(static_cast<const StructArray&>(array), i, rep_level);
+  }
+
+  Status Handle(const StructArray& array, int64_t i, int16_t rep_level) {
+    if (array.IsNull(i)) {
+      RETURN_NOT_OK(AppendLevelsRecursive(rep_level, int16_t(def_level_ - 1)));
+    } else {
+      RETURN_NOT_OK(rep_levels_builder_->Append(rep_level));
+      RETURN_NOT_OK(def_levels_builder_->Append(def_level_));
+      for (int j = 0; j < num_children(); j++) {
+        RETURN_NOT_OK(children_[j]->Handle(*array.field(j), i, rep_level));
+      }
+    }
+    return Status::OK();
+  }
+
+  Status Finish(ColumnMap* colmap) override {
+    std::shared_ptr<Array> rep_levels_tmp;
+    std::shared_ptr<Array> def_levels_tmp;
+    RETURN_NOT_OK(rep_levels_builder_->Finish(&rep_levels_tmp));
+    RETURN_NOT_OK(def_levels_builder_->Finish(&def_levels_tmp));
+    colmap->Put(field_, std::static_pointer_cast<Int16Array>(rep_levels_tmp),
+                        std::static_pointer_cast<Int16Array>(def_levels_tmp),
+                        nullptr);
+    return Status::OK();
+  }
+};
+
+class ListShredNode : public ShredNode {
+ public:
+  using ShredNode::ShredNode;
+
+  static Status Create(const std::shared_ptr<Field>& field,
+                     int16_t rep_level, int16_t def_level,
+                     MemoryPool* pool, std::unique_ptr<ShredNode>* node_out) {
+    if (field->type()->num_children() != 1) {
+      return Status::Invalid("List fields must have exactly one child: ", field->name());
+    }
+    std::unique_ptr<ListShredNode> node(new ListShredNode(field, pool));
+    node->rep_level_ = ++rep_level;
+    if (field->nullable()) {
+      def_level++;
+    }
+    def_level++; // empty list
+    node->def_level_ = def_level;
+    std::unique_ptr<ShredNode> child;
+    RETURN_NOT_OK(CreateTree(field->type()->child(0), rep_level, def_level, pool, &child));
+    node->children_.push_back(std::move(child));
+    *node_out = std::move(node);
+    return Status::OK();
+  }
+
+  Status Handle(const Array& array, int64_t i, int16_t rep_level) override {
+    return Handle(static_cast<const ListArray&>(array), i, rep_level);
```

Comment:
you might want to use checked_cast here.

Reply:
Done.